ext/json/ext/parser/parser.c

```diff
@@ -857,32 +859,89 @@ static const bool string_scan[256] = {
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
 };
 
+#if (defined(__GNUC__ ) || defined(__clang__))
+#define FORCE_INLINE __attribute__((always_inline))
+#else
+#define FORCE_INLINE
+#endif
+
+static inline bool FORCE_INLINE string_scan_basic(JSON_ParserState *state)
+{
+    while (state->cursor < state->end) {
+        if (RB_UNLIKELY(string_scan[(unsigned char)*state->cursor])) {
+            return 1;
+        }
+        *state->cursor++;
+    }
+    return 0;
+}
+
+static bool (*string_scan_impl)(JSON_ParserState *);
+
+#ifdef HAVE_SIMD
+#ifdef HAVE_SIMD_NEON
+
+static inline FORCE_INLINE bool string_scan_neon(JSON_ParserState *state)
+{
+    while (state->cursor + sizeof(uint8x16_t) <= state->end) {
+        uint8x16_t chunk = vld1q_u8((const unsigned char *)state->cursor);
+
+        // Trick: c < 32 || c == 34 can be factored as c ^ 2 < 33
+        // https://lemire.me/blog/2025/04/13/detect-control-characters-quotes-and-backslashes-efficiently-using-swar/
+        const uint8x16_t too_low_or_dbl_quote = vcltq_u8(veorq_u8(chunk, vdupq_n_u8(2)), vdupq_n_u8(33));
+
+        uint8x16_t has_backslash = vceqq_u8(chunk, vdupq_n_u8('\\'));
+        uint8x16_t needs_escape  = vorrq_u8(too_low_or_dbl_quote, has_backslash);
+
+        // Benchmarking this on an M1 Macbook Air shows that this is faster than
+        // using the neon_match_mask function (see the generator.c code) and bit
+        // operations to find the first match.
+        if (vmaxvq_u8(needs_escape)) {
```

Comment:
I assume this is why you didn't share the between the two? Because I'd much prefer share it for maintainability reasons.